<!-- dm_arm_bringup.launch -->

<launch>
    <!-- ==================== 参数 ==================== -->
    <arg name="use_rviz" default="true"/>
    <arg name="use_fake_execution" default="false" doc="是否使用假执行模式"/>
    <arg name="config_file" default="$(find dm_arm_service)/config/dm_arm_config.yaml"/>
    
    <!-- ==================== 加载统一配置 ==================== -->
    <rosparam file="$(arg config_file)" command="load"/>
    
    <!-- ==================== 机器人描述 ==================== -->
    <!-- <param name="robot_description" 
           textfile="$(find dm_ht_controller)/urdf/dm_arm.urdf"/> -->
    <param name="robot_description"
            textfile="$(find dm_arm_description)/urdf/DM-Arm-Description.urdf"/>

    <!-- ==================== 状态发布器 ==================== -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" 
          type="robot_state_publisher" output="screen"/>
    
    <!-- ==================== 硬件接口（仅在真实模式下启动） ==================== -->
    <group unless="$(arg use_fake_execution)">
        <node name="dm_controller" pkg="dm_ht_controller" 
              type="dm_controller" output="screen" 
              required="true">
        </node>
        
        <node name="controller_spawner" pkg="controller_manager" type="spawner" 
              output="screen" respawn="false"
              args="joint_state_controller arm_controller end_controller"/>
    </group>
    
    <!-- ==================== 假执行模式（仅在假执行模式下启动） ==================== -->
    <group if="$(arg use_fake_execution)">
        <node name="joint_state_publisher" pkg="joint_state_publisher" 
              type="joint_state_publisher" output="screen">
            <param name="use_gui" value="false"/>
            <rosparam param="source_list">
                [/move_group/fake_controller_joint_states]
            </rosparam>
        </node>
    </group>
    
    <!-- ==================== MoveIt ==================== -->
    <include file="$(find dm_arm_moveit_config)/launch/move_group.launch">
        <arg name="allow_trajectory_execution" value="true"/>
        <arg name="info" value="true"/>
    </include>
    
    <!-- ==================== 服务器 ==================== -->
    <node name="dm_arm_server" pkg="dm_arm_service" type="dm_arm_server" 
          output="screen" respawn="true"
          launch-prefix="bash -c 'sleep 3; $0 $@'">
    </node>
    
    <!-- ==================== RViz ==================== -->
    <node name="rviz" pkg="rviz" type="rviz" if="$(arg use_rviz)"
          args="-d $(find dm_arm_moveit_config)/launch/moveit.rviz"
          launch-prefix="bash -c 'sleep 5; $0 $@'"/>

</launch>
